### PR TITLE
experiments: Taint propagators

### PR DESCRIPTION
**Pull request by IagoAbal**
Closes PA-1684

# Thanks for improving Semgrep Docs 😀

### Please ensure

- [x] A subject matter expert (SME) reviews the content
- [x] A technical writer reviews the content or PR
- [x] This change has no security implications or else you have pinged the security team
- [x] Redirects are added if the PR changes page URLs

### CODE REVIEW

File: docs/experiments/taint-propagators.md
Comment (adamkvitek) [inline docs/experiments/taint-propagators.md:35]: @IagoAbal what do we mean by propagation through assignments?

```suggestion
Semgrep only propagates taint through assignments. As a consequence, 
Semgrep cannot find the issue using the rule displayed above.
```

File: docs/experiments/taint-propagators.md
Comment (adamkvitek) [inline docs/experiments/taint-propagators.md:38]:
```suggestion
The use of **taint propagators** enables Semgrep to propagate taint in these scenarios.
Taint propagators are specified under the `pattern-propagators` key:
```

File: docs/experiments/taint-propagators.md
Comment (adamkvitek) [inline docs/experiments/taint-propagators.md:41]: @IagoAbal I was a bit confused by naming - how do we want to call this feature? Is it:

- Pattern propagator (as mentioned in the code snippet here)
- Taint propagator (as in the header)
- Propagator (as we call it in some paragraphs)

We can find one name and stick to it.

File: docs/experiments/taint-propagators.md
Comment (adamkvitek) [inline docs/experiments/taint-propagators.md:51]:
```suggestion
In the example above, whenever Semgrep finds the pattern `$X.add($Y)`, it checks if
`$Y` is tainted. If `$Y` is tainted, Semgrep propagates that taint to `$X`. Thus,
adding tainted data to a set marks the set itself as tainted.

<iframe src="https://semgrep.dev/embed/editor?snippet=7lNe" border="0" frameBorder="0" width="100%" height="432"></iframe>

## Propagator rule requirements

A propagator rule necessitates:
- Any arbitrary pattern that must bind two metavariables of your choice.
- You must specify the `from` and the `to` metavariables.
```

File: docs/experiments/taint-propagators.md
Comment (adamkvitek) [inline docs/experiments/taint-propagators.md:53]: Thank you for this example! @IagoAbal (Just moving it above.)

Comment (adamkvitek) [reply docs/experiments/taint-propagators.md:53]:
```suggestion
```

**Review by adamkvitek — APPROVED** [review]
@IagoAbal thank you! I made some suggestions, please check them. Approving so you are not blocked, but if my comments require further discussion, feel free to wait before merging (leaving this up to you) - we can always fix anything by future PRs.

File: docs/experiments/taint-propagators.md
Comment (IagoAbal) [reply docs/experiments/taint-propagators.md:22]: Why "database injection"? This is a generic "injection" vulnerability, I'm just using a generic `sink` method.

File: docs/experiments/taint-propagators.md
Comment (IagoAbal) [reply docs/experiments/taint-propagators.md:41]: The feature is "taint propagators" but the operator is `pattern-propagators` because all our operators start with `pattern-`. That we cannot change. Now, "propagator" is just a short form of "taint propagator". We could rewrite those if you insist.

File: docs/experiments/taint-propagators.md
Comment (IagoAbal) [reply docs/experiments/taint-propagators.md:9]: I don't like this header much because here the spec and the example are intermixed. If we had a more formal spec of what a propagator is, then followed by an example, then it would make more sense to me.

File: docs/experiments/taint-propagators.md
Comment (adamkvitek) [reply docs/experiments/taint-propagators.md:45]: @IagoAbal thanks for the explanation, looks like you fixed it!

**Pull request merged**
